### PR TITLE
Rig: Bottleneck animation management through a Rig class

### PR DESCRIPTION
**Pull request by howard-stearns**
The Rig will eventually manage avatar/NPC activity state (walk, turn, idle, etc.) and combine the various multiple animations and overrides (e.g., controllers). Right now, it's just trying to bring all the appropriate information together without breaking anything. That means completely scripted bone animation (such as walk.js) should still work (although when the Rig work is completed, walk.js would not be needed). Also scripts that start and stop animations through MyAvatar should still work.

There is also  an animation test suite, with a single test case called Rig. It currently doesn't do much, but there's doc in Rig.h. 

Also incorporates "remove unused legacy Shapes from Model and friends" (https://github.com/howard-stearns/hifi/pull/8)

This is version is not meant as the final API or class structure, merely one that illustrates the concept while maintaining the above. All the heavy lifting for bottlenecking was by Seth, and there's more comming. (But enough here to get the ball rolling on merging.)

The only new behavior is a very crude state machine within the Rig, that applies the first "idle" role animation found for your avatar, and toggles between that and the first "walk" role animation found. It does this ONLY when the 'enableRig' setting is true. (All the bottlenecking of control happens regardless, but the automated start/stop of idle/walk happens only when enableRig is true.) For example:
var idleAnimation = "https://hifi-public.s3.amazonaws.com/ozan/support/FightClubBotTest1/Animations/standard_idle.fbx"
var walkAnimation = "https://hifi-public.s3.amazonaws.com/ozan/support/FightClubBotTest1/Animations/standard_walk.fbx";
// NB: Testing is being done with avatar https://hifi-public.s3.amazonaws.com/ozan/support/FightClubBotTest1/FightClubBot.fst
MyAvatar.startAnimationByRole("idle", idleAnimation, 30, 1, true, false); // Idles in place
MyAvatar.stopAnimation(idleAnimation);
MyAvatar.startAnimationByRole("walk", walkAnimation, 30, 1, true, false, 0, 60);  // Walks in place
MyAvatar.stopAnimation(walkAnimation);
Settings.setValue("enableRig", true);  // And then restart Interface to have it take effect. Set to false and restart to cancel.
// All of this, including the animations, will be retained on restart.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1bb734aec0bbf6e1eb4e0b64627832307d6d8e16 available at: https://gist.github.com/2a3372fdcfaed80dd8c2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.dmg

File: interface/src/avatar/SkeletonModel.cpp [hunk 1]
```diff
@@ -151,48 +175,13 @@ void SkeletonModel::simulate(float deltaTime, bool fullUpdate) {
         }
     }
 
-    if (_isFirstPerson) {
-        cauterizeHead();
+    // if (_isFirstPerson) {
+    //     cauterizeHead();
+    //     updateClusterMatrices();
+    // }
```
Comment (AndrewMeadows) [inline interface/src/avatar/SkeletonModel.cpp]: This cruft should probably be deleted outright: the **cauterizeHead()** method was deleted.

File: interface/src/Camera.h [hunk 2]
```diff
@@ -28,7 +28,7 @@ enum CameraMode
 };
 
 Q_DECLARE_METATYPE(CameraMode);
-static int cameraModeId = qRegisterMetaType<CameraMode>();
+// static int cameraModeId = qRegisterMetaType<CameraMode>();
```
Comment (ZappoMan) [inline interface/src/Camera.h]: I believe this will break some scripts.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4754615159f5c4f554181d59e67cc367db2177c2 available at: https://gist.github.com/2a3372fdcfaed80dd8c2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.dmg

File: interface/src/Camera.h (hunk 2 above)
Comment (howard-stearns) [inline interface/src/Camera.h]: @sethalves Do we need to change Camera.h? What's the relationship?

Comment (sethalves) [inline interface/src/Camera.h]: I don't remember... probably trying to quiet the compiler?  let's just put it back.

Comment (ZappoMan) [inline interface/src/Camera.h]: This exposes the data type to JS. We should keep it.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 711dce4edc99f510269c789b8816eab233bfdb50 available at: https://gist.github.com/2a3372fdcfaed80dd8c2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 12d449f69d92a8eb85a5e9f4966fbd7b9acd23a1 available at: https://gist.github.com/2a3372fdcfaed80dd8c2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6c68dbeee26605e05d94c87d892ccaf8565b1daf available at: https://gist.github.com/2a3372fdcfaed80dd8c2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1dddf18243e1fb773980eb27c8dffeeaf4e3f12a available at: https://gist.github.com/2a3372fdcfaed80dd8c2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ebd0e38d4972e582f0d8122ed6ac6b9cc8777b62 available at: https://gist.github.com/2a3372fdcfaed80dd8c2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 95c43c4857d1a63b5aa940306c860b7c26b1aec9 available at: https://gist.github.com/2a3372fdcfaed80dd8c2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 06aa1077622e19ab5dd2ab3fdfff0a8c780b4496 available at: https://gist.github.com/2a3372fdcfaed80dd8c2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5e3bebcd8679501ceef12e721660486b76c1e2ca available at: https://gist.github.com/2a3372fdcfaed80dd8c2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.dmg

**Comment by howard-stearns** [comment]
There are still two bugs, but it's "pencils down" time. Let's merge this thing!
https://app.asana.com/0/32622044445063/43174289852667
https://app.asana.com/0/26225263936266/43174289852671

**Comment by hifi-gustavo** [comment]
Build console output for commit: 31574521c1a282a7146e12bae08c444477f171b9 available at: https://gist.github.com/2a3372fdcfaed80dd8c2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6ad3cdd4027a378accfda5ee852dcafc0be9d40f available at: https://gist.github.com/d019bcf19f554071e781<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.dmg

**Comment by howard-stearns** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6ad3cdd4027a378accfda5ee852dcafc0be9d40f available at: https://gist.github.com/2a3372fdcfaed80dd8c2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 40abd63b61702fdaea590a1dc484e84a04f33615 available at: https://gist.github.com/2a3372fdcfaed80dd8c2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 51ed912f5e0417616f83372f0ca8c6e9fe7aefa7 available at: https://gist.github.com/2a3372fdcfaed80dd8c2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b685378ab01bdb23c507850f112fcf0ab489ffc8 available at: https://gist.github.com/2a3372fdcfaed80dd8c2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.dmg

**Comment by howard-stearns** [comment]
Blocker! Bad Access in the following when (enough? specific?) other avatars are present:

0   io.highfidelity.Interface     	0x0000000106d1e680 Model::initJointTransforms() + 240
1   io.highfidelity.Interface     	0x0000000106d1e4d4 Model::setScale(glm::detail::tvec3<float, (glm::precision)0> const&) + 164
2   io.highfidelity.Interface     	0x000000010691cd8d SkeletonModel::simulate(float, bool) + 429
3   io.highfidelity.Interface     	0x00000001068e140c Avatar::simulate(float) + 1036
4   io.highfidelity.Interface     	0x00000001068f1060 AvatarManager::simulateAvatarFades(float) + 544
5   io.highfidelity.Interface     	0x00000001068f0c85 AvatarManager::updateOtherAvatars(float) + 885

**Comment by howard-stearns** [comment]
I cannot reliably reproduce the Bad Access, but I've committed a speculative fix.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 271311533682ed67acd591271709dd76d79d077a available at: https://gist.github.com/2a3372fdcfaed80dd8c2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5438.dmg

File: libraries/animation/src/Rig.cpp [hunk 3]
```diff
@@ -0,0 +1,754 @@
...
+    }
+
+    // now update the joint states from the top
+    for (int j = freeLineage.size() - 1; j >= 0; j--) {
+        updateJointState(freeLineage.at(j), parentTransform);
+    }
+
+    return true;
+}
+
```
Comment (AndrewMeadows) [inline libraries/animation/src/Rig.cpp]: A historical note:  the code above was originally in Model and used some IK logic that had a few problems (awkward solutions, as I recall, but maybe some instabilities).  I rewrote it and provided the inverseKinematics() code below to try to solve some of the specific problems but did not rip out the old stuff -- using it was still an option at the time.

Comment (howard-stearns) [inline libraries/animation/src/Rig.cpp]: Yup. SkeletonModel::applyPalmData still has the three-way options in this branch: MenuOption::AlternateIK, MenuOption::AlignForearmsWithWrists, and the "standard" inverseKinematics,

**Comment by thoys** [comment]
@howard-stearns @AndrewMeadows ,  this PR broke all the ModelOverlays , please see the lobby.js . The problem is that _rig is set as a nullptr on creation of the Model object instance of the ModelOverlay.

**Comment by thoys** [comment]
fixed the issue in https://github.com/highfidelity/hifi/pull/5493

**Pull request merged**